Mask GKE Sandbox-specific labels and taints

impl opened the pull request:
Hi friends,

This change prevents the GKE Sandbox labels and taints (currently `sandbox.gke.io/runtime=gvisor`) from causing node pools to always need to be recreated. Without this change, once a node pool is created, it needs to be manually updated to include the additional GKE Sandbox label and taint items.

I believe this PR fixes https://github.com/terraform-providers/terraform-provider-google/issues/4210.

The addition of the `DiffSuppressFunc` on a `TypeList` is a little weird (not sure if it's using prescribed functionality or not; I see https://github.com/hashicorp/terraform-plugin-sdk/issues/477 but it seems to work here). I'd be happy to take any suggestions on how to improve it.

Once I get an OK on the general approach, I will add a few tests as well. Thanks for taking a look!

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
    
Unless you choose release-note:none, please add a release note.

See .ci/RELEASE_NOTES_GUIDE.md for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: Masked automatically applied GKE Sandbox node labels and taints on node pools
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.

Thanks for your contribution!  A human will be with you soon.

@danawillow, please review this PR or find an appropriate assignee.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3749-old..auto-pr-3749) ( 1 file changed, 64 insertions(+), 4 deletions(-))

danawillow commented on third_party/terraform/utils/node_config.go.erb:
```diff
@@ -502,4 +509,58 @@ func flattenSandboxConfig(c *containerBeta.SandboxConfig) []map[string]interface
 }
+
+func containerNodePoolLabelsSuppress(k, old, new string, d *schema.ResourceData) bool {
+	o, n := d.GetChange("node_config.0.sandbox_config.0.sandbox_type")
```
I believe this is trying to suggest that we only care about the sandbox label if sandbox_type is set, is that correct? If so can you add a comment to that regard?

impl replied:
This is indeed the case; I added a comment about this to both methods.

danawillow commented on third_party/terraform/utils/node_config.go.erb:
```diff
@@ -502,4 +509,58 @@ func flattenSandboxConfig(c *containerBeta.SandboxConfig) []map[string]interface
+		return false
+	}
+
+	if strings.HasPrefix(k, "node_config.0.labels.sandbox.gke.io/") {
```
I'm not convinced this will work. It would be worth verifying (with some print statements or a test) but I'm pretty sure 'k' here is going to be the full list, rather than this func being called on each element of the list.

impl replied:
So the behavior here is really interesting. You actually get each key and value under the root independently and can return a different result for each. For example, dumping the `k`, `old`, and `new` values for each invocation you'd see something like:

```
k=node_config.0.labels.# old=2 new=1
k=node_config.0.labels.test.terraform.io/test old=foo new=foo
k=node_config.0.labels.test.sandbox.gke.io/runtime old=gvisor new=
```

It turns out to not be important (see my next comment) now, but the initial pass did depend on this admittedly strange behavior. I think it has to do with `old` and `new` being strings.

danawillow commented on third_party/terraform/utils/node_config.go.erb:
```diff
@@ -502,4 +509,58 @@ func flattenSandboxConfig(c *containerBeta.SandboxConfig) []map[string]interface
 	return result
 }
+
+func containerNodePoolLabelsSuppress(k, old, new string, d *schema.ResourceData) bool {
```
the node config schema gets used both in the node pool resource and in the container cluster resource (and within container cluster, it's top level and part of the node pool block), so the address of the field you're looking for will be different depending on where the user is specifying it. I'd recommend using the value of 'k' to help you figure out where you are.

impl replied:
I think they both actually use the same hierarchy by chance, but it raises the well-taken point that there's no restriction on this schema being used elsewhere in the future. I have updated the code to derive the root of the node config from `k` and added a comment to clarify.

danawillow commented on third_party/terraform/utils/node_config.go.erb:
```diff
@@ -502,4 +509,58 @@ func flattenSandboxConfig(c *containerBeta.SandboxConfig) []map[string]interface
+	return false
+}
+
+func containerNodePoolTaintSuppress(k, old, new string, d *schema.ResourceData) bool {
```
I think we have a fairly similar function in resource_container_cluster: containerClusterAddedScopesSuppress. Can you look at that and see if using similar logic makes sense here?

impl replied:
This logic makes _way_ more sense than the initial way I put this PR together. I have updated the code to use a similar strategy. Note that it does get called multiple times -- once for each "stringable" key/value Terraform provides -- but I think the performance hit is worth the clarity in this case. Now the suppression will be consistent across the entire field, rather than changing for some sub-keys (it is not clear at all what the upward propagation behavior is). Thank you for the pointer!

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3749-old..auto-pr-3749) ( 3 files changed, 274 insertions(+), 4 deletions(-))

impl commented:
> Here are some comments for a first pass! I'd definitely like to see some tests so we can be sure this actually works.

Thank you for the quick and detailed review! I've updated the code to address your concerns and added tests for both the `container_cluster` and `container_node_pool` resources.

danawillow commented:
Hey, just wanted to pop in and let you know I haven't forgotten about this! I've had a bunch of other things on my plate, but I should be able to look at this again next week.

impl commented:
Hi there, no worries! Totally understand. Thanks for the update -- much appreciated!

danawillow commented on third_party/terraform/tests/resource_container_cluster_test.go.erb:
```diff
@@ -890,6 +890,26 @@ func TestAccContainerCluster_withSandboxConfig(t *testing.T) {
+				// GKE sets automatic labels and taints on nodes. This makes
+				// sure we ignore the automatic ones and keep our own.
+				Config:   testAccContainerCluster_withSandboxConfig(clusterName),
+				PlanOnly: true,
```
Can you also add a comment here about what this does, since it's a less common feature in our tests?

impl replied:
Done!

danawillow commented on third_party/terraform/tests/resource_container_cluster_test.go.erb:
```diff
@@ -890,6 +890,26 @@ func TestAccContainerCluster_withSandboxConfig(t *testing.T) {
+				PlanOnly: true,
+			},
+			{
+				// Now we'll modify the labels, which should force a change to
```
FYI labels aren't modifiable in-place, so this test step destroys and recreates the cluster. If that's fine for this test, can you add a note here so it doesn't surprise people if they come across this?

impl replied:
This is indeed intentional, and I've added a comment to note that!

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3749-old..auto-pr-3749) ( 3 files changed, 286 insertions(+), 4 deletions(-))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3749-old..auto-pr-3749) ( 3 files changed, 286 insertions(+), 4 deletions(-))
